### ISSUE TITLE
Shipping-information endpoint returns error when DiscountData is specified in the body

### ISSUE THREAD

**Issue by apedicdev** [magento/magento2#30780]
### Preconditions (*)

1. Reproduced on **2.4-develop**
2. Replicated on 2.3.5-p1 and 2.4.1

### Steps to reproduce (*)

1. `POST` request to `rest/default/V1/guest-carts/{cartId}/shipping-information`
Request body from `https://magento.redoc.ly/2.4.0-admin/tag/guest-cartscartIdshipping-information`
```
{
  "addressInformation": {
    "shipping_address": {
      "id": 0,
      "region": "string",
      "region_id": 0,
      "region_code": "string",
      "country_id": "string",
      "street": [
        "string"
      ],
      "company": "string",
      "telephone": "string",
      "fax": "string",
      "postcode": "string",
      "city": "string",
      "firstname": "string",
      "lastname": "string",
      "middlename": "string",
      "prefix": "string",
      "suffix": "string",
      "vat_id": "string",
      "customer_id": 0,
      "email": "string",
      "same_as_billing": 0,
      "customer_address_id": 0,
      "save_in_address_book": 0,
      "extension_attributes": {
        "discounts": [
          {
            "discount_data": {
              "amount": 0,
              "base_amount": 0,
              "original_amount": 0,
              "base_original_amount": 0
            },
            "rule_label": "string",
            "rule_id": 0
          }
        ],
        "gift_registry_id": 0,
        "pickup_location_code": "string"
      },
      "custom_attributes": [
        {
          "attribute_code": "string",
          "value": "string"
        }
      ]
    },
    "billing_address": {
      "id": 0,
      "region": "string",
      "region_id": 0,
      "region_code": "string",
      "country_id": "string",
      "street": [
        "string"
      ],
      "company": "string",
      "telephone": "string",
      "fax": "string",
      "postcode": "string",
      "city": "string",
      "firstname": "string",
      "lastname": "string",
      "middlename": "string",
      "prefix": "string",
      "suffix": "string",
      "vat_id": "string",
      "customer_id": 0,
      "email": "string",
      "same_as_billing": 0,
      "customer_address_id": 0,
      "save_in_address_book": 0,
      "extension_attributes": {
        "discounts": [
          {
            "discount_data": {
              "amount": 0,
              "base_amount": 0,
              "original_amount": 0,
              "base_original_amount": 0
            },
            "rule_label": "string",
            "rule_id": 0
          }
        ],
        "gift_registry_id": 0,
        "pickup_location_code": "string"
      },
      "custom_attributes": [
        {
          "attribute_code": "string",
          "value": "string"
        }
      ]
    },
    "shipping_method_code": "string",
    "shipping_carrier_code": "string",
    "extension_attributes": {},
    "custom_attributes": [
      {
        "attribute_code": "string",
        "value": "string"
      }
    ]
  }
}
```

### Expected result (*)

Improved error handling, the API is supposed to warn the client that those fields are read-only and are not suppose to be sent on the POST request.

### Actual result (*)
```
{
    "message": "Property \"DiscountData\" does not have accessor method \"setDiscountData\" in class \"Magento\\SalesRule\\Api\\Data\\RuleDiscountInterface\".",
    "trace": "#0 /bitnami/magento/htdocs/vendor/magento/framework/Reflection/NameFinder.php(77): Magento\\Framework\\Reflection\\NameFinder->findAccessorMethodName(Object(Laminas\\Code\\Reflection\\ClassReflection), 'DiscountData', 'setDiscountData', 'setIsDiscountDa...')\n#1 /bitnami/magento/htdocs/vendor/magento/framework/Webapi/ServiceInputProcessor.php(265): Magento\\Framework\\Reflection\\NameFinder->getSetterMethodName(Object(Laminas\\Code\\Reflection\\ClassReflection), 'DiscountData')\n#2 /bitnami/magento/htdocs/vendor/magento/framework/Webapi/ServiceInputProcessor.php(474): Magento\\Framework\\Webapi\\ServiceInputProcessor->_createFromArray('\\\\Magento\\\\SalesR...', Array)\n#3 /bitnami/magento/htdocs/vendor/magento/framework/Webapi/ServiceInputProcessor.php(277): Magento\\Framework\\Webapi\\ServiceInputProcessor->convertValue(Array, '\\\\Magento\\\\SalesR...')\n#4 /bitnami/magento/htdocs/vendor/magento/framework/Webapi/ServiceInputProcessor.php(478): Magento\\Framework\\Webapi\\ServiceInputProcessor->_createFromArray('\\\\Magento\\\\Quote\\\\...', Array)\n#5 /bitnami/magento/htdocs/vendor/magento/framework/Webapi/ServiceInputProcessor.php(277): Magento\\Framework\\Webapi\\ServiceInputProcessor->convertValue(Array, '\\\\Magento\\\\Quote\\\\...')\n#6 /bitnami/magento/htdocs/vendor/magento/framework/Webapi/ServiceInputProcessor.php(478): Magento\\Framework\\Webapi\\ServiceInputProcessor->_createFromArray('\\\\Magento\\\\Quote\\\\...', Array)\n#7 /bitnami/magento/htdocs/vendor/magento/framework/Webapi/ServiceInputProcessor.php(277): Magento\\Framework\\Webapi\\ServiceInputProcessor->convertValue(Array, '\\\\Magento\\\\Quote\\\\...')\n#8 /bitnami/magento/htdocs/vendor/magento/framework/Webapi/ServiceInputProcessor.php(478): Magento\\Framework\\Webapi\\ServiceInputProcessor->_createFromArray('Magento\\\\Checkou...', Array)\n#9 /bitnami/magento/htdocs/vendor/magento/framework/Webapi/ServiceInputProcessor.php(165): Magento\\Framework\\Webapi\\ServiceInputProcessor->convertValue(Array, 'Magento\\\\Checkou...')\n#10 /bitnami/magento/htdocs/vendor/magento/module-webapi/Controller/Rest/InputParamsResolver.php(84): Magento\\Framework\\Webapi\\ServiceInputProcessor->process('Magento\\\\Checkou...', 'saveAddressInfo...', Array)\n#11 /bitnami/magento/htdocs/vendor/magento/framework/Interception/Interceptor.php(58): Magento\\Webapi\\Controller\\Rest\\InputParamsResolver->resolve()\n#12 /bitnami/magento/htdocs/vendor/magento/framework/Interception/Interceptor.php(138): Magento\\Webapi\\Controller\\Rest\\InputParamsResolver\\Interceptor->___callParent('resolve', Array)\n#13 /bitnami/magento/htdocs/vendor/magento/framework/Interception/Interceptor.php(153): Magento\\Webapi\\Controller\\Rest\\InputParamsResolver\\Interceptor->Magento\\Framework\\Interception\\{closure}()\n#14 /bitnami/magento/htdocs/generated/code/Magento/Webapi/Controller/Rest/InputParamsResolver/Interceptor.php(23): Magento\\Webapi\\Controller\\Rest\\InputParamsResolver\\Interceptor->___callPlugins('resolve', Array, Array)\n#15 /bitnami/magento/htdocs/vendor/magento/module-webapi/Controller/Rest/SynchronousRequestProcessor.php(85): Magento\\Webapi\\Controller\\Rest\\InputParamsResolver\\Interceptor->resolve()\n#16 /bitnami/magento/htdocs/vendor/magento/module-webapi/Controller/Rest.php(188): Magento\\Webapi\\Controller\\Rest\\SynchronousRequestProcessor->process(Object(Magento\\Framework\\Webapi\\Rest\\Request\\Proxy))\n#17 /bitnami/magento/htdocs/vendor/magento/framework/Interception/Interceptor.php(58): Magento\\Webapi\\Controller\\Rest->dispatch(Object(Magento\\Framework\\App\\Request\\Http))\n#18 /bitnami/magento/htdocs/vendor/magento/framework/Interception/Interceptor.php(138): Magento\\Webapi\\Controller\\Rest\\Interceptor->___callParent('dispatch', Array)\n#19 /bitnami/magento/htdocs/vendor/magento/framework/Interception/Interceptor.php(153): Magento\\Webapi\\Controller\\Rest\\Interceptor->Magento\\Framework\\Interception\\{closure}(Object(Magento\\Framework\\App\\Request\\Http))\n#20 /bitnami/magento/htdocs/generated/code/Magento/Webapi/Controller/Rest/Interceptor.php(23): Magento\\Webapi\\Controller\\Rest\\Interceptor->___callPlugins('dispatch', Array, Array)\n#21 /bitnami/magento/htdocs/vendor/magento/framework/App/Http.php(116): Magento\\Webapi\\Controller\\Rest\\Interceptor->dispatch(Object(Magento\\Framework\\App\\Request\\Http))\n#22 /bitnami/magento/htdocs/generated/code/Magento/Framework/App/Http/Interceptor.php(23): Magento\\Framework\\App\\Http->launch()\n#23 /bitnami/magento/htdocs/vendor/magento/framework/App/Bootstrap.php(263): Magento\\Framework\\App\\Http\\Interceptor->launch()\n#24 /bitnami/magento/htdocs/index.php(39): Magento\\Framework\\App\\Bootstrap->run(Object(Magento\\Framework\\App\\Http\\Interceptor))\n#25 {main}"
}
```
See in the **exception.log** file:

![Screenshot from 2020-11-04 16-31-34](https://user-images.githubusercontent.com/51679138/98250042-a5eccf80-1f7f-11eb-9fb3-a41edb242238.png)

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by floorz** [comment]
@apedicdev I started typing out a really long investigative message for what I've discovered about this issue, but might have found the cause and solution. You don't happen to use Avalara for tax / address validation do you?

Check this out: https://github.com/avadev/Avalara-AvaTax-for-Magento2/releases/tag/2.1.9

![image](https://user-images.githubusercontent.com/4982916/98194417-9b2d2e80-1edc-11eb-9a4f-0627d5070040.png)

I've been having the same exact error for a client of mine since we upgraded to Magento 2.4.0.

- As part of the address validation of the Avalara Tax module, it fetches quote address data which gets the "discount" extension attribute populated on it if a customer has applied a coupon code.
- The extension attributes mistakingly get set with the billing address and sent in the place order request when trying to complete checkout.
- Magento's web api processor uses Interfaces to determine which data in a webapi request is valid to set in the Interfaces implementing Model as well as computing the php "setters" to call to set that data on the model. In this case it sees "discount_data", so it tries to call `\Magento\SalesRule\Api\Data::setDiscountData()` which doesn't exist.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Alfa 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-38924` were created

**Issue Available**: @engcom-Alfa, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by apedicdev** [comment]
@magento give me 2.4-develop instance

**Comment by apedicdev** [comment]
@floorz this has been replicated on magento clean installation using payload from documentation `https://magento.redoc.ly/2.4.0-admin/tag/guest-cartscartIdshipping-information` 
I guess if the `discount_data` isn't needed, then documentation should be fixed accordingly. Viceversa, setters need to be implemented. @engcom-Alfa can you please confirm what is the default behaviour you want?

**Comment by apedicdev** [comment]
I can confirm issue has been replicated on vanilla Magento instance

Steps to replicate:

1. https://2e8070d80c43c244cae2951e56b914c5-2-4-develop.instances.magento-community.engineering
2. add a product to cart as a guest user and proceed to checkout
3. Copy cart-id assigned from any XHR request
3. `POST` to `https://2e8070d80c43c244cae2951e56b914c5-2-4-develop.instances.magento-community.engineering/rest/default/V1/guest-carts/{cart-id}/shipping-information`

Payload from https://magento.redoc.ly/2.4.0-admin/tag/guest-cartscartIdshipping-information
```
{
  "addressInformation": {
    "shipping_address": {
      "id": 0,
      "region": "string",
      "region_id": 0,
      "region_code": "string",
      "country_id": "string",
      "street": [
        "string"
      ],
      "company": "string",
      "telephone": "string",
      "fax": "string",
      "postcode": "string",
      "city": "string",
      "firstname": "string",
      "lastname": "string",
      "middlename": "string",
      "prefix": "string",
      "suffix": "string",
      "vat_id": "string",
      "customer_id": 0,
      "email": "string",
      "same_as_billing": 0,
      "customer_address_id": 0,
      "save_in_address_book": 0,
      "extension_attributes": {
        "discounts": [
          {
            "discount_data": {
              "amount": 0,
              "base_amount": 0,
              "original_amount": 0,
              "base_original_amount": 0
            },
            "rule_label": "string",
            "rule_id": 0
          }
        ],
        "gift_registry_id": 0,
        "pickup_location_code": "string"
      },
      "custom_attributes": [
        {
          "attribute_code": "string",
          "value": "string"
        }
      ]
    },
    "billing_address": {
      "id": 0,
      "region": "string",
      "region_id": 0,
      "region_code": "string",
      "country_id": "string",
      "street": [
        "string"
      ],
      "company": "string",
      "telephone": "string",
      "fax": "string",
      "postcode": "string",
      "city": "string",
      "firstname": "string",
      "lastname": "string",
      "middlename": "string",
      "prefix": "string",
      "suffix": "string",
      "vat_id": "string",
      "customer_id": 0,
      "email": "string",
      "same_as_billing": 0,
      "customer_address_id": 0,
      "save_in_address_book": 0,
      "extension_attributes": {
        "discounts": [
          {
            "discount_data": {
              "amount": 0,
              "base_amount": 0,
              "original_amount": 0,
              "base_original_amount": 0
            },
            "rule_label": "string",
            "rule_id": 0
          }
        ],
        "gift_registry_id": 0,
        "pickup_location_code": "string"
      },
      "custom_attributes": [
        {
          "attribute_code": "string",
          "value": "string"
        }
      ]
    },
    "shipping_method_code": "string",
    "shipping_carrier_code": "string",
    "extension_attributes": {},
    "custom_attributes": [
      {
        "attribute_code": "string",
        "value": "string"
      }
    ]
  }
}
```
**Result Expected:**
```
{
  "payment_methods": [
    {
      "code": "string",
      "title": "string"
    }
  ],
  "totals": {
    "grand_total": 0,
    "base_grand_total": 0,
    "subtotal": 0,
    "base_subtotal": 0,
    "discount_amount": 0,
    "base_discount_amount": 0,
    "subtotal_with_discount": 0,
    "base_subtotal_with_discount": 0,
    "shipping_amount": 0,
    "base_shipping_amount": 0,
    "shipping_discount_amount": 0,
    "base_shipping_discount_amount": 0,
    "tax_amount": 0,
    "base_tax_amount": 0,
    "weee_tax_applied_amount": 0,
    "shipping_tax_amount": 0,
    "base_shipping_tax_amount": 0,
    "subtotal_incl_tax": 0,
    "base_subtotal_incl_tax": 0,
    "shipping_incl_tax": 0,
    "base_shipping_incl_tax": 0,
    "base_currency_code": "string",
    "quote_currency_code": "string",
    "coupon_code": "string",
    "items_qty": 0,
    "items": [
      {
        "item_id": 0,
        "price": 0,
        "base_price": 0,
        "qty": 0,
        "row_total": 0,
        "base_row_total": 0,
        "row_total_with_discount": 0,
        "tax_amount": 0,
        "base_tax_amount": 0,
        "tax_percent": 0,
        "discount_amount": 0,
        "base_discount_amount": 0,
        "discount_percent": 0,
        "price_incl_tax": 0,
        "base_price_incl_tax": 0,
        "row_total_incl_tax": 0,
        "base_row_total_incl_tax": 0,
        "options": "string",
        "weee_tax_applied_amount": 0,
        "weee_tax_applied": "string",
        "extension_attributes": {
          "negotiable_quote_item_totals": {
            "cost": 0,
            "catalog_price": 0,
            "base_catalog_price": 0,
            "catalog_price_incl_tax": 0,
            "base_catalog_price_incl_tax": 0,
            "cart_price": 0,
            "base_cart_price": 0,
            "cart_tax": 0,
            "base_cart_tax": 0,
            "cart_price_incl_tax": 0,
            "base_cart_price_incl_tax": 0,
            "extension_attributes": {}
          }
        },
        "name": "string"
      }
    ],
    "total_segments": [
      {
        "code": "string",
        "title": "string",
        "value": 0,
        "area": "string",
        "extension_attributes": {
          "tax_grandtotal_details": [
            {
              "amount": 0,
              "rates": [
                {
                  "percent": "string",
                  "title": "string"
                }
              ],
              "group_id": 0
            }
          ],
          "gift_cards": "string",
          "gw_order_id": "string",
          "gw_item_ids": [
            "string"
          ],
          "gw_allow_gift_receipt": "string",
          "gw_add_card": "string",
          "gw_price": "string",
          "gw_base_price": "string",
          "gw_items_price": "string",
          "gw_items_base_price": "string",
          "gw_card_price": "string",
          "gw_card_base_price": "string",
          "gw_base_tax_amount": "string",
          "gw_tax_amount": "string",
          "gw_items_base_tax_amount": "string",
          "gw_items_tax_amount": "string",
          "gw_card_base_tax_amount": "string",
          "gw_card_tax_amount": "string",
          "gw_price_incl_tax": "string",
          "gw_base_price_incl_tax": "string",
          "gw_card_price_incl_tax": "string",
          "gw_card_base_price_incl_tax": "string",
          "gw_items_price_incl_tax": "string",
          "gw_items_base_price_incl_tax": "string",
          "vertex_tax_calculation_messages": [
            "string"
          ]
        }
      }
    ],
    "extension_attributes": {
      "coupon_label": "string",
      "negotiable_quote_totals": {
        "items_count": 0,
        "quote_status": "string",
        "created_at": "string",
        "updated_at": "string",
        "customer_group": 0,
        "base_to_quote_rate": 0,
        "cost_total": 0,
        "base_cost_total": 0,
        "original_total": 0,
        "base_original_total": 0,
        "original_tax": 0,
        "base_original_tax": 0,
        "original_price_incl_tax": 0,
        "base_original_price_incl_tax": 0,
        "negotiated_price_type": 0,
        "negotiated_price_value": 0
      },
      "base_customer_balance_amount": 0,
      "customer_balance_amount": 0,
      "reward_points_balance": 0,
      "reward_currency_amount": 0,
      "base_reward_currency_amount": 0
    }
  },
  "extension_attributes": {}
}
```

**Actual result:**
`"message": "Internal Error. Details are available in Magento log file. Report ID: webapi-5fa5225d54019`

**Comment by apedicdev** [comment]
@magento I am working on this

**Comment by apedicdev** [comment]
@magento add to contributors team

**Comment by apedicdev** [comment]
@magento I am working on this

**Comment by apedicdev** [comment]
this has been replicated on magento clean installation using payload from documentation `https://magento.redoc.ly/2.4.0-admin/tag/guest-cartscartIdshipping-information`.
I guess if the discount_data isn't needed, then documentation should be fixed accordingly. Viceversa, setters need to be implemented. @engcom-Alfa @magento-engcom-team can you please confirm what should be the default behaviour?

**Comment by magento-engcom-team** [comment]
Hi @apedicdev. 

Thank you for your report and collaboration!

The related internal Jira ticket `MC-38924` was closed as `Fixed`. 

The fix will be available with the upcoming `2.4.3` release.

**Comment by sramesh1988** [comment]
Hi 
Can you please share the PR link or commit id.
We are facing this issue in 2.4.2-P1
